docs/storage_plugin_development.md

```diff
@@ -1,55 +1,143 @@
+
+  // ReadOnlyInterface's virtual overrides here
+};
+
+// Implementations
+
+}  // namespace my_namespace
+
+// The following block exposes our class to pluginlib so that it can be discovered at runtime.
 #include "pluginlib/class_list_macros.hpp"
-PLUGINLIB_EXPORT_CLASS(MyReadonlyStorage, rosbag2_storage::storage_interfaces::ReadOnlyInterface)
+PLUGINLIB_EXPORT_CLASS(my_namespace::MyReadOnlyStorage,
+                       rosbag2_storage::storage_interfaces::ReadOnlynterface)
 ```
-with the plugin description
+
 ```
-<library path="my_storage_lib">
-  <class name="my_readonly_storage" type="MyReadonlyStorage" base_class_type="rosbag2_storage::storage_interfaces::ReadOnlyInterface">
+<!-- plugin_description.xml -->
+<library path="rosbag2_storage_my_storage">
+  <class
+    name="my_readonly_storage"
+    type="my_namespace::MyReadOnlyStorage"
+    base_class_type="rosbag2_storage::storage_interfaces::ReadOnlyInterface"
+  >
+    <description>Rosbag2 storage plugin providing read functionality for MyStorage file format.</description>
   </class>
 </library>
 ```
+
 and the usual pluginlib export in the CMakeLists:
 
-`pluginlib_export_plugin_description_file(rosbag2_storage plugin_description.xml)`
+```
+# CMakeLists.txt
+pluginlib_export_plugin_description_file(rosbag2_storage plugin_description.xml)
+```
+
+## Providing plugin-specific configuration
+
+Some storage plugins may have configuration parameters unique to the format that you'd like to allow users to provide from the command line.
+Rosbag2 provides a CLI argument `--storage-config-file` which allows users to pass the path to a file.
+This file can contain anything, its format is specified by the storage implementation, it is passed as a path all the way to the plugin, where it may be used however desired.
+Plugins are recommended to document the expected format of this file so that users can write well-formatted configurations.
+
+### Extending CLI from a storage plugin
+
+Commandline arguments can be a much more convenient way to expose configuration to users than writing out a file.
+The `ros2bag` package, which creates the `ros2 bag` command, provides an entrypoint for plugins to extend the CLI.
+
+All a package needs to do is expose a Python setuptools entrypoint to the group `ros2bag.storage_plugin_cli_extension`, with an entrypoint keyed by the name of the storage plugin. For example, here is `setup.cfg` from `rosbag2_storage_mcap`:
+
+```
+[options.entry_points]
+ros2bag.storage_plugin_cli_extension =
+  mcap = ros2bag_mcap_cli
+```
+
+This registers an entrypoint in group `ros2bag.storage_plugin_cli_extension`, for the plugin named `mcap`, that is implemented by a Python module called `rosbag2_mcap_cli`.
+
+The exposed entrypoint can be installed as a Python module by any method, for example via `ament_cmake_python`'s `ament_python_install_package` macro, or by having a pure-python `ament_python` package with a `setup.py`.
+
+The functions this entrypoint may provide:
```

Comment:
would be good to include what happens by default if this entrypoint exists but does not expose any functions.